### ISSUE TITLE
Keep worker stateless

### ISSUE THREAD

**Issue by QiJune** [sql-machine-learning/elasticdl#1253]
Stateless worker is good for elastic scheduling. 

Training/Validation data information is packed into a task. Worker gets tasks from master to do training or evaluation. If some worker is failed, we could just ignore the task.

In my mind, task related status is no need to keep. This will waste some mini-batched data, but  keep the logic of master and worker clean.

**Comment by mhaoli** [comment]
Ignoring some of the tasks is fine.
However, it should be noted that workers may be killed frequently if there are insufficient cluster resources or problems with the cluster machines. This can cause too many tasks to be ignored.
As far as I know, there are indeed machine learning systems that ignore failed tasks. But when the number of ignored tasks reaches a certain user-defined percentage of the total number of tasks, the entire training job fails. Users can start a new job to train from some checkpoints.
ElasticDL shouldn't let the entire task fail, and we can see if there is any better way.

**Comment by QiJune** [comment]
The training dataset could be very large, and it also will be trained for lots of epochs. 
So, I am not sure if some tasks failing will hurt the targets: time-to-coverage or accuracy-to-coverage.

**Comment by mhaoli** [comment]
In my experience, some people indeed train for only 1 epoch because the training dataset is too large. I guess this may be common in the situation of online learning.

Also, if there are some problems with the machine the worker pod is running on, the worker may fail a lot of tasks. The percentage of failed task may be the indicator whether the whole training system is running correctly.

**Comment by skydoorkai** [comment]
This will change the "fault-tolerance" of ElasticDL. A worker may get multiple tasks in its hand.

**Comment by chunyang-wen** [comment]
This can be an option. But we should not trade correctness for clearance. Previously users have requested a feature like this:
> If they use ten workers to train their model, the job finishes when
> + chief worker finishes
> + a limited number of workers fail or hang
>
> the job fails when
> + chief worker fails
> + two many worker fails

It is similar. We can provide reasonable options like what you suggested.

Besides, whether a worker can be stateless is determined by many things such as the operator state of a TensorFlow op which is not controlled by elasticdl. Machine learning is robust, but we should respect the actual training process.

**Comment by brightcoder01** [comment]
The worker will report each task is done successfully or with failure to master. If the task contains N mini-batches, only M(M < N) mini-batches are processed and then the failure happens, then the whole task will be recovered and the whole N mini-batches will be processes by another worker (M mini-batches will be processed twice). It's acceptable because the entire dataset is large.

Worker don't keep any state, all the task states is kept in the Task Queue in master. What worker do is fetching task, processing task and reporting task done.
